cou/steps/execute.py

```diff
@@ -25,6 +26,66 @@
 logger = logging.getLogger(__name__)
 
 
+async def _run_sub_stpes_in_parallel(
+    step: BaseStep, prompt: bool, overwrite_progress: bool
+) -> None:
+    """Run all sub-steps of step in parallel.
+
+    :param step: Step to be executed.
+    :type step: BaseStep
+    :param prompt: Whether to run upgrade step with prompt (interactive mode).
+    :type prompt: bool
+    :param overwrite_progress: Whether to overwrite the progress indication message in CLI output
+                               for all sub-steps. True to overwrite and False (the default) to
+                               persist.
+    :type overwrite_progress: bool
+    :raises HaltUpgradeExecution: When any step raised it. All other steps will be completed.
+    """
+    logger.debug("running all sub-steps of %s step in parallel", step)
+    grouped_coroutines = (
+        apply_step(sub_step, prompt, overwrite_progress) for sub_step in step.sub_steps
+    )
+    try:
+        await asyncio.gather(*grouped_coroutines)
```

Comment:
nit: Is the default value, but I think this makes a little bit more explicit why the other awaitables won't be cancelled
```suggestion
        await asyncio.gather(*grouped_coroutines, return_exceptions=False)
```

Reply:
Even if `return_exceptions=True` others will not be canceled. So I do not see using default value here. 
If we use `return_exceptions=True` it will mean that exception will be returned and not raised, so we can collect result of any coroutine.

Reply:
Interesting....
I've played around to see the behavior and I think we should use `return_exceptions=True` to be sure that all tasks are completed. See the following example

```python

import asyncio

async def fn1(x):
    await asyncio.sleep(3)
    print(f"fn1 is called with x={x}")
    return f"Hello from fn1: {x}"
    
async def raise_error(y):
    await asyncio.sleep(2)  # raise error is faster than fn1
    print(f"raise_error is called with y={y}")
    raise asyncio.TimeoutError()
```

See that like this the first and third call does not print
```python
result1 = await asyncio.gather(fn1("First call"),raise_error("Second call"), fn1("Third call"), return_exceptions=False)

> raise_error is called with y=Second call
> TimeoutError                              Traceback (most recent call last)
```

If we to a try except and give more time the other tasks are completed:
```python
try:
    result1 = await asyncio.gather(fn1("First call"),raise_error("Second call"), fn1("Third call"), return_exceptions=False)

except (asyncio.TimeoutError):
    await asyncio.sleep(10)
    raise
    
> raise_error is called with y=Second call
> fn1 is called with x=First call
> fn1 is called with x=Third call
```
But adding this sleep is dirty, and when change `return_exceptions` it works as expected:
```
result1 = await asyncio.gather(fn1("First call"),raise_error("Second call"), fn1("Third call"), return_exceptions=True)

> result1
['Hello from fn1: First call',
 asyncio.exceptions.TimeoutError(),
 'Hello from fn1: Third call']
```

Reply:
You are right, I did not consider that task will be finished, but we're already handling the exception -> going to next task.
I think that at the end we will not to do `return_exceptions=True` and check if any return value is based on Exception. All value should be None only those raising exception will be some kind of exception, but like this we will not exit if another value is raised, e.g. `ValueError`. That's not perfect, either. Alternatively, we can wait for all steps to finished in the exception section and then raise the exception.
What do you think?

Reply:
> , but like this we will not exit if another value is raised, e.g. ValueError

I didn't understand this phrase. If more than one task raises an exception, I think it's a good thing to have it in the results and then compile to the user which tasks failed

> Alternatively, we can wait for all steps to finished in the exception section and then raise the exception.

What if another exception happens while waiting the others to finish it? I don't think its a good approach